client/battle/BattleFieldController.cpp

```diff
@@ -333,6 +360,129 @@ std::set<BattleHex> BattleFieldController::getHighlightedHexesForMovementTarget(
+
+	// overwrite full ranged damage distance from Additional info field of LIMITED_SHOOTING_RANGE bonus
+	auto bonus = hoveredStack->getBonus(Selector::type()(BonusType::LIMITED_SHOOTING_RANGE));
+	if(bonus != nullptr && bonus->additionalInfo != CAddInfo::NONE)
+		rangedFullDamageDistance = bonus->additionalInfo[0];
+
+	// get only battlefield hexes that are in full range damage distance
+	std::set<BattleHex> fullRangeLimit; 
+	for(auto & hex : battleFieldWithoutSideColumns)
+	{
+		if(BattleHex::getDistance(hoveredHex, hex) <= rangedFullDamageDistance)
+			rangedFullDamageHexes.push_back(hex);
+	}
+
+	return rangedFullDamageHexes;
+}
+
+std::vector<BattleHex> BattleFieldController::getRangedFullDamageLimitHexes(std::vector<BattleHex> rangedFullDamageHexes)
+{
+	std::vector<BattleHex> rangedFullDamageLimitHexes; // used for return
+
+	// if not a hovered arcer unit -> return
+	auto hoveredHex = getHoveredHex();
+	const CStack * hoveredStack = owner.curInt->cb->battleGetStackByPos(hoveredHex, true);
+	if(!(hoveredStack && hoveredStack->isShooter()))
+		return rangedFullDamageLimitHexes;
+
+	auto rangedFullDamageDistance = GameConstants::BATTLE_PENALTY_DISTANCE;
+
+	// overwrite full ranged damage distance from Additional info field of LIMITED_SHOOTING_RANGE bonus
+	auto bonus = hoveredStack->getBonus(Selector::type()(BonusType::LIMITED_SHOOTING_RANGE));
+	if(bonus != nullptr && bonus->additionalInfo != CAddInfo::NONE)
+		rangedFullDamageDistance = bonus->additionalInfo[0];
+
+	// from ranged full damage hexes get only the ones at the limit
+	for(auto & hex : rangedFullDamageHexes)
+	{
+		if(BattleHex::getDistance(hoveredHex, hex) == rangedFullDamageDistance)
+			rangedFullDamageLimitHexes.push_back(hex);
+	}
+
+	return rangedFullDamageLimitHexes;
+}
+
+std::vector<std::vector<BattleHex::EDir>> BattleFieldController::getOutsideNeighbourDirectionsForLimitHexes(std::vector<BattleHex> rangedFullDamageHexes, std::vector<BattleHex> rangedFullDamageLimitHexes)
+{
+	std::vector<std::vector<BattleHex::EDir>> output;
+
+	if(rangedFullDamageHexes.empty())
+		return output;
+
+	for(auto & hex : rangedFullDamageLimitHexes)
+	{
+		// get all neighbours and their directions
+		auto neighbours = hex.neighbouringTilesWithDirection();
+
+		std::vector<BattleHex::EDir> outsideNeighbourDirections;
+
+		// for each neighbour add to output only the ones not found in rangedFullDamageHexes
+		for(auto & neighbour : neighbours)
+		{
+			auto it = std::find(rangedFullDamageHexes.begin(), rangedFullDamageHexes.end(), neighbour.second);
+
+			if(it == rangedFullDamageHexes.end())
+				outsideNeighbourDirections.push_back(neighbour.first); // push direction
+		}
+
+		output.push_back(outsideNeighbourDirections);
+	}
+
+	return output;
+}
+
+// krs: This needs to be moved somewhere else
+unsigned char setBit(unsigned char mask, int position)
+{
+    if (position >= 0 && position < 8)
+        mask |= (1 << position);
+
+    return mask;
+}

```

Comment:
just use std::bitset

Reply:
@Nordsoft91 to use a bit set in serializable environments, we need to merge #1822

Reply:
This option does not needs serialization, so bitset is OK (and probably makes sense here)

Reply:
If does not need - then it can be used. But I think than #1822 will beerged before this.

Reply:
changed to:
```
uint8_t setBit(std::bitset<6> mask, int position)
{
    if (position >= 0 && position < 6)
        mask.set(position);

    return static_cast<uint8_t>(mask.to_ulong());
}
```

Reply:
You don't need this function at all. Use bitset as type of "mask" variable that you use with this function

Reply:
but I need a byte here:
`output.push_back(fullDamageRangeLimitImages[mask]);`

Reply:
ok changed to this:
```
	for(auto & directions : fullRangedDamageLimitHexesNeighbourDirections)
	{
		std::bitset<6> mask;
		
		// convert directions to mask
		for(auto direction : directions)
			mask.set(direction);

		uint8_t imageKey = static_cast<uint8_t>(mask.to_ulong());
		output.push_back(fullDamageRangeLimitImages[imageKey]);
	}
```

Reply:
What I mean is that in `calculateFullRangedDamageHighlightImages` method you should change
`uint8_t mask = 0;`
to
`std::bitset<6> mask`

And call `mask.set(direction)` instead of current `mask = setBit(mask, direction);`

And once you need numeric value convert bitset to integer via 
`output.push_back(fullDamageRangeLimitImages[mask.to_ulong()]);`

Reply:
Ah, yeah, correct :)